// @trademe/ng-defer-load/esm5/defer-load.module.js
import * as tslib_1 from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { DeferLoadDirective } from './defer-load.directive';
import * as ɵngcc0 from '@angular/core';
var DeferLoadModule = /** @class */ (function () {
    function DeferLoadModule() {
    }
DeferLoadModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: DeferLoadModule });
DeferLoadModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function DeferLoadModule_Factory(t) { return new (t || DeferLoadModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DeferLoadModule, { declarations: function () { return [DeferLoadDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [DeferLoadDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DeferLoadModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [DeferLoadDirective],
                exports: [DeferLoadDirective]
            }]
    }], function () { return []; }, null); })();
    return DeferLoadModule;
}());
export { DeferLoadModule };
