// ng-vertical-timeline/esm5/lib/ng-vertical-timeline.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { NgVerticalTimelineComponent } from './ng-vertical-timeline.component';
import * as ɵngcc0 from '@angular/core';
var NgVerticalTimelineModule = /** @class */ (function () {
    function NgVerticalTimelineModule() {
    }
NgVerticalTimelineModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgVerticalTimelineModule });
NgVerticalTimelineModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgVerticalTimelineModule_Factory(t) { return new (t || NgVerticalTimelineModule)(); }, imports: [[BrowserModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgVerticalTimelineModule, { declarations: function () { return [NgVerticalTimelineComponent]; }, imports: function () { return [BrowserModule]; }, exports: function () { return [NgVerticalTimelineComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgVerticalTimelineModule, [{
        type: NgModule,
        args: [{
                imports: [BrowserModule],
                declarations: [NgVerticalTimelineComponent],
                exports: [NgVerticalTimelineComponent]
            }]
    }], function () { return []; }, null); })();
    return NgVerticalTimelineModule;
}());
export { NgVerticalTimelineModule };
