// ng-vertical-timeline/bundles/ng-vertical-timeline.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),require('@angular/platform-browser'),exports, require('@angular/core'), require('@angular/platform-browser')) :
    typeof define === 'function' && define.amd ? define('ng-vertical-timeline', ['@angular/core','@angular/common','@angular/platform-browser','exports', '@angular/core', '@angular/platform-browser'], factory) :
    (factory(global.ng.core,global.ng.common,global.ng.platformBrowser,(global['ng-vertical-timeline'] = {}),global.ng.core,global.ng.platformBrowser));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,exports,i0,platformBrowser) { 
function NgVerticalTimelineComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "div", 2);
    ɵngcc0.ɵɵelementStart(2, "div", 3);
    ɵngcc0.ɵɵelementStart(3, "div", 4);
    ɵngcc0.ɵɵelementStart(4, "div", 5);
    ɵngcc0.ɵɵelement(5, "img", 6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "div", 7);
    ɵngcc0.ɵɵelementStart(7, "span");
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(10, "div", 8);
    ɵngcc0.ɵɵelement(11, "p", 9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var item_r1 = ctx.$implicit;
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r0.isPair(item_r1.id) ? "left" : "right");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵstyleProp("background-color", ctx_r0.primaryColor);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("src", item_r1.icon, ɵngcc0.ɵɵsanitizeUrl);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(item_r1.date);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", item_r1.title, " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("innerHTML", item_r1.content, ɵngcc0.ɵɵsanitizeHtml);
} }
'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgVerticalTimelineService = /** @class */ (function () {
        function NgVerticalTimelineService() {
        }
        /** @nocollapse */
        NgVerticalTimelineService.ctorParameters = function () { return []; };
        /** @nocollapse */ NgVerticalTimelineService.ngInjectableDef = i0.defineInjectable({ factory: function NgVerticalTimelineService_Factory() { return new NgVerticalTimelineService(); }, token: NgVerticalTimelineService, providedIn: "root" });
NgVerticalTimelineService.ɵfac = function NgVerticalTimelineService_Factory(t) { return new (t || NgVerticalTimelineService)(); };
NgVerticalTimelineService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: NgVerticalTimelineService, factory: function (t) { return NgVerticalTimelineService.ɵfac(t); }, providedIn: 'root' });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgVerticalTimelineService, [{
        type: i0.Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return []; }, null); })();
        return NgVerticalTimelineService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgVerticalTimelineComponent = /** @class */ (function () {
        function NgVerticalTimelineComponent() {
        }
        /**
         * @return {?}
         */
        NgVerticalTimelineComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () { };
        /**
         * @param {?} id
         * @return {?}
         */
        NgVerticalTimelineComponent.prototype.isPair = /**
         * @param {?} id
         * @return {?}
         */
            function (id) {
                return !(id % 2);
            };
        /** @nocollapse */
        NgVerticalTimelineComponent.ctorParameters = function () { return []; };
        NgVerticalTimelineComponent.propDecorators = {
            data: [{ type: i0.Input }],
            primaryColor: [{ type: i0.Input }]
        };
NgVerticalTimelineComponent.ɵfac = function NgVerticalTimelineComponent_Factory(t) { return new (t || NgVerticalTimelineComponent)(); };
NgVerticalTimelineComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NgVerticalTimelineComponent, selectors: [["ng-vertical-timeline"]], inputs: { data: "data", primaryColor: "primaryColor" }, decls: 2, vars: 1, consts: [[1, "timeline"], [4, "ngFor", "ngForOf"], [1, "container", "left", 3, "ngClass"], [1, "inner"], [1, "head"], [1, "icon"], [1, "ui", "tiny", "image", 3, "src"], [1, "title"], [1, "body"], [3, "innerHTML"]], template: function NgVerticalTimelineComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, NgVerticalTimelineComponent_div_1_Template, 12, 7, "div", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.data);
    } }, directives: [ɵngcc1.NgForOf, ɵngcc1.NgClass], styles: [".timeline[_ngcontent-%COMP%]{position:relative;max-width:1200px;margin:0 auto}.timeline[_ngcontent-%COMP%]:after{content:\"\";position:absolute;width:6px;background-color:#222;top:0;bottom:0;left:50%;margin-left:-3px}.timeline[_ngcontent-%COMP%]   .container[_ngcontent-%COMP%]{position:relative;width:50%;margin:initial;padding:10px 40px;background-color:initial;box-sizing:border-box}.timeline[_ngcontent-%COMP%]   .container[_ngcontent-%COMP%]:after{content:\"\";position:absolute;width:25px;height:25px;right:-16px;background-color:#000428;border:4px solid #004e92;top:15px;border-radius:50%;z-index:1}.timeline[_ngcontent-%COMP%]   .container[_ngcontent-%COMP%]   .inner[_ngcontent-%COMP%]{background-color:#fff;color:#222;box-shadow:0 3px 6px rgba(0,0,0,.16),0 3px 6px rgba(0,0,0,.23)}.timeline[_ngcontent-%COMP%]   .container[_ngcontent-%COMP%]   .inner[_ngcontent-%COMP%]   .head[_ngcontent-%COMP%]{display:flex;align-items:center;color:#fff;background-color:#222}.timeline[_ngcontent-%COMP%]   .container[_ngcontent-%COMP%]   .inner[_ngcontent-%COMP%]   .head[_ngcontent-%COMP%]   .icon[_ngcontent-%COMP%]{margin:20px;max-width:50px}.timeline[_ngcontent-%COMP%]   .container[_ngcontent-%COMP%]   .inner[_ngcontent-%COMP%]   .head[_ngcontent-%COMP%]   .icon[_ngcontent-%COMP%]   img[_ngcontent-%COMP%]{width:100%;height:100%}.timeline[_ngcontent-%COMP%]   .container[_ngcontent-%COMP%]   .inner[_ngcontent-%COMP%]   .head[_ngcontent-%COMP%]   .title[_ngcontent-%COMP%]{font-size:24px;text-transform:uppercase}.timeline[_ngcontent-%COMP%]   .container[_ngcontent-%COMP%]   .inner[_ngcontent-%COMP%]   .head[_ngcontent-%COMP%]   .title[_ngcontent-%COMP%]   span[_ngcontent-%COMP%]{display:block;font-size:14px}.timeline[_ngcontent-%COMP%]   .container[_ngcontent-%COMP%]   .inner[_ngcontent-%COMP%]   .body[_ngcontent-%COMP%]{background-color:#fff;padding:20px}.left[_ngcontent-%COMP%]{left:0}.left[_ngcontent-%COMP%]:before{content:\" \";height:0;position:absolute;top:22px;width:0;z-index:1;right:30px;border:solid #222;border-width:10px 0 10px 10px;border-color:transparent transparent transparent #222}.right[_ngcontent-%COMP%]{left:50%}.right[_ngcontent-%COMP%]:before{content:\" \";height:0;position:absolute;top:22px;width:0;z-index:1;left:30px;border:solid #222;border-width:10px 10px 10px 0;border-color:transparent #222 transparent transparent}.right[_ngcontent-%COMP%]:after{left:-16px}@media screen and (max-width:600px){.timeline[_ngcontent-%COMP%]:after{left:31px}.timeline[_ngcontent-%COMP%]   .container[_ngcontent-%COMP%]{width:100%;padding-left:70px;padding-right:25px}.timeline[_ngcontent-%COMP%]   .container[_ngcontent-%COMP%]:before{left:60px;border:solid #222;border-width:10px 10px 10px 0;border-color:transparent #222 transparent transparent}.left[_ngcontent-%COMP%]::after, .right[_ngcontent-%COMP%]::after{left:12px}.right[_ngcontent-%COMP%]{left:0}#timeline--1[_ngcontent-%COMP%], #timeline--2[_ngcontent-%COMP%], #timeline--3[_ngcontent-%COMP%]{top:initial}}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgVerticalTimelineComponent, [{
        type: i0.Component,
        args: [{
                selector: 'ng-vertical-timeline',
                template: "<div class=\"timeline\">\r\n\r\n  <div *ngFor=\"let item of data\">\r\n\r\n    <div class=\"container left\" [ngClass]=\"isPair(item.id)?'left':'right'\">\r\n      <div class=\"inner\">\r\n        <div class=\"head\" [style.backgroundColor]=\"primaryColor\">\r\n          <div class=\"icon\">\r\n            <img class=\"ui tiny image\" [src]=\"item.icon\">\r\n          </div>\r\n\r\n          <div class=\"title\">\r\n            <span>{{item.date}}</span>\r\n            {{item.title }}\r\n          </div>\r\n\r\n        </div>\r\n        <div class=\"body\">\r\n          <p [innerHTML]=\"item.content\"></p>\r\n        </div>\r\n      </div>\r\n    </div>\r\n  </div>\r\n</div>\r\n",
                styles: [".timeline{position:relative;max-width:1200px;margin:0 auto}.timeline:after{content:\"\";position:absolute;width:6px;background-color:#222;top:0;bottom:0;left:50%;margin-left:-3px}.timeline .container{position:relative;width:50%;margin:initial;padding:10px 40px;background-color:initial;box-sizing:border-box}.timeline .container:after{content:\"\";position:absolute;width:25px;height:25px;right:-16px;background-color:#000428;border:4px solid #004e92;top:15px;border-radius:50%;z-index:1}.timeline .container .inner{background-color:#fff;color:#222;box-shadow:0 3px 6px rgba(0,0,0,.16),0 3px 6px rgba(0,0,0,.23)}.timeline .container .inner .head{display:flex;align-items:center;color:#fff;background-color:#222}.timeline .container .inner .head .icon{margin:20px;max-width:50px}.timeline .container .inner .head .icon img{width:100%;height:100%}.timeline .container .inner .head .title{font-size:24px;text-transform:uppercase}.timeline .container .inner .head .title span{display:block;font-size:14px}.timeline .container .inner .body{background-color:#fff;padding:20px}.left{left:0}.left:before{content:\" \";height:0;position:absolute;top:22px;width:0;z-index:1;right:30px;border:solid #222;border-width:10px 0 10px 10px;border-color:transparent transparent transparent #222}.right{left:50%}.right:before{content:\" \";height:0;position:absolute;top:22px;width:0;z-index:1;left:30px;border:solid #222;border-width:10px 10px 10px 0;border-color:transparent #222 transparent transparent}.right:after{left:-16px}@media screen and (max-width:600px){.timeline:after{left:31px}.timeline .container{width:100%;padding-left:70px;padding-right:25px}.timeline .container:before{left:60px;border:solid #222;border-width:10px 10px 10px 0;border-color:transparent #222 transparent transparent}.left::after,.right::after{left:12px}.right{left:0}#timeline--1,#timeline--2,#timeline--3{top:initial}}"]
            }]
    }], function () { return []; }, { data: [{
            type: i0.Input
        }], primaryColor: [{
            type: i0.Input
        }] }); })();
        return NgVerticalTimelineComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgVerticalTimelineModule = /** @class */ (function () {
        function NgVerticalTimelineModule() {
        }
NgVerticalTimelineModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgVerticalTimelineModule });
NgVerticalTimelineModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgVerticalTimelineModule_Factory(t) { return new (t || NgVerticalTimelineModule)(); }, imports: [[platformBrowser.BrowserModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgVerticalTimelineModule, { declarations: [NgVerticalTimelineComponent], imports: [ɵngcc2.BrowserModule], exports: [NgVerticalTimelineComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgVerticalTimelineModule, [{
        type: i0.NgModule,
        args: [{
                imports: [platformBrowser.BrowserModule],
                declarations: [NgVerticalTimelineComponent],
                exports: [NgVerticalTimelineComponent]
            }]
    }], function () { return []; }, null); })();
        return NgVerticalTimelineModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.NgVerticalTimelineService = NgVerticalTimelineService;
    exports.NgVerticalTimelineComponent = NgVerticalTimelineComponent;
    exports.NgVerticalTimelineModule = NgVerticalTimelineModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
